File: sdk/typescript/src/webhook-secret.test.ts
import { describe, expect, it } from 'vitest';
import { createHmac } from 'node:crypto';
import { BanzamiWebhookSignatureError, constructEvent, verifySignature } from './webhooks';

// A2-03. An integration that never configured its webhook secret verified
// against "" — so an event signed with the empty key was accepted, and a forged
// payment_link.paid shipped goods. An empty or blank secret now refuses.
describe('webhook verification needs a secret', () => {
  const body = JSON.stringify({ id: 'evt_1', type: 'payment_link.paid', data: {} });
  const t = Math.floor(Date.now() / 1000);
  const forgedWithEmptyKey = `t=${t},v1=${createHmac('sha256', '').update(`${t}.${body}`).digest('hex')}`;

  for (const secret of ['', '   ']) {
    it(`refuses to verify with secret ${JSON.stringify(secret)}`, () => {
      expect(() => verifySignature(body, forgedWithEmptyKey, secret)).toThrow(BanzamiWebhookSignatureError);
      expect(() => constructEvent(body, forgedWithEmptyKey, secret)).toThrow(BanzamiWebhookSignatureError);
    });
  }
});

File: sdk/go/banzami/webhook_secret_test.go
package banzami

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/hex"
	"fmt"
	"testing"
	"time"
)

// A2-03. An integration that never configured its secret verified against "",
// so an event signed with the empty key was accepted. It is now refused.
func TestVerifySignature_RefusesAnEmptySecret(t *testing.T) {
	body := []byte(`{"id":"evt_1","type":"payment_link.paid","data":{}}`)
	ts := time.Now().Unix()
	mac := hmac.New(sha256.New, []byte(""))
	fmt.Fprintf(mac, "%d.", ts)
	mac.Write(body)
	forged := fmt.Sprintf("t=%d,v1=%s", ts, hex.EncodeToString(mac.Sum(nil)))
	for _, secret := range []string{"", "   "} {
		if err := VerifySignature(body, forged, secret, 5*time.Minute); err == nil {
			t.Fatalf("secret %q: an event signed with the empty key was accepted", secret)
		}
	}
}
